src/include/migraphx/op/convolution.hpp

```diff
@@ -69,43 +69,99 @@ struct convolution
         }
 
-        const shape& input   = inputs.at(0);
-        const shape& weights = inputs.at(1);
-        size_t kdims         = input_size - 2;
-        if(kdims != this->kdims())
+        const shape& input            = inputs.at(0);
+        const shape& weights          = inputs.at(1);
+        const size_t num_spatial_dims = input_size - 2;
+        if(num_spatial_dims != this->kdims())
         {
-            MIGRAPHX_THROW("convolution: input k-dims does not match attribute size");
+            MIGRAPHX_THROW("CONVOLUTION: input k-dims does not match attribute size");
         }
 
-        if(input.lens().at(1) != (weights.lens().at(1) * group))
-            MIGRAPHX_THROW("CONVOLUTION: Mismatch channel numbers");
-
-        std::vector<size_t> output_lens{input.lens()[0], weights.lens()[0]};
-
-        for(size_t i = 0; i < kdims; i++)
+        if(not input.dynamic() and not weights.dynamic() and
+           input.lens().at(1) != (weights.lens().at(1) * group))
+            MIGRAPHX_THROW("CONVOLUTION: mismatched channel numbers");
+
+        auto calc_output_lens = [&](std::vector<std::size_t> i_lens,
+                                    std::vector<std::size_t> w_lens) {
+            std::vector<size_t> ret = {};
+            // calculate the output shape of the convolution: ((W - K + 2P) / S) + 1
+            for(size_t i = 0; i < num_spatial_dims; i++)
+            {
+                if(i_lens.at(i) == 0 or w_lens.at(i) == 0)
+                {
+                    // for handling when a dimension = 0 (opt of dynamic_dimension)
+                    ret.push_back(0);
+                }
+                else
+                {
+                    auto padding_factor = 2 * padding[i];
+                    if(padding_size == 2 * num_spatial_dims)
+                    {
+                        // when padding is {x0_begin, x1_begin, ... x0_end , x1_end, ...}
+                        padding_factor = padding[i] + padding[i + num_spatial_dims];
+                    }
+                    ret.push_back(std::size_t(std::max<std::ptrdiff_t>(
+                        1,
+                        (i_lens[i + 2] - (1 + dilation[i] * (w_lens[i + 2] - 1)) + padding_factor) /
+                                stride[i] +
+                            1)));
+                }
+            }
+            return ret;
+        };
+
+        if(input.dynamic() or weights.dynamic())
         {
-            auto padding_factor = 2 * padding[i];
-            if(padding_size == 2 * kdims)
-                padding_factor = padding[i] + padding[i + kdims];
-            output_lens.push_back(std::size_t(std::max<std::ptrdiff_t>(
-                1,
-                (input.lens()[i + 2] - (1 + dilation[i] * (weights.lens()[i + 2] - 1)) +
-                 padding_factor) /
-                        stride[i] +
-                    1)));
+            std::vector<shape::dynamic_dimension> output_dyn_dims = {};
```

Comment:
currently if neither the input nor the weights are dynamic, we are populating this vector anyways. I suggest just calling it output_dims.

Reply:
That should not be the case. If both of the input shapes are fixed, the `output_lens` in the else block should be used.

Reply:
yes, I saw the if statement before. But my second comment on the else block is asking if we need to do that or we can merge both into one calculation.